Fix for Issue #62, Issue #67

elebitzero opened the pull request:
- Issue #62
Added a custom horizontal offset for all characters(including previous event characters) to center their faces on the 10 pull portrait. 

After the fix, Kaeya and Razor are looking better: 
![image](https://user-images.githubusercontent.com/42903164/107333547-7c55af80-6a6a-11eb-8d9d-e1eb7cef7a93.png)
![image](https://user-images.githubusercontent.com/42903164/107333737-bb840080-6a6a-11eb-9df7-895dc63cb8be.png)

- Issue #67
Restores the currently selected banner after wishing so it doesn't start from Beginner's wish and starts with the current banner like in the game.

elebitzero commented:
I also modified the code temporarily to get specific characters so we could see a before/after:
![image](https://user-images.githubusercontent.com/42903164/107334288-5da3e880-6a6b-11eb-9518-963f2989c7a1.png)
![image](https://user-images.githubusercontent.com/42903164/107334327-6d233180-6a6b-11eb-9faf-84fb951b7ba8.png)

uzair-ashraf commented:
@elebitzero so I love the idea, and it looks really good, however, have you tested this on mobile? Using percentage based widths on safari get kinda wonky, and the containers can get smaller leading to the offset being weird. I think the best thing to do would be modifying the character images in photoshop to bring them more to center.

Let me know how it looks on mobile!

elebitzero commented:
I tested the app on Android and it looks good there as well:

Before:
![image](https://user-images.githubusercontent.com/42903164/107470698-a8336c80-6b20-11eb-8032-d7c2353f4ed8.png)
After:
![image](https://user-images.githubusercontent.com/42903164/107470730-b5505b80-6b20-11eb-8290-e3db54d06534.png)

elebitzero commented:
> Using percentage based widths on safari get kinda wonky

I'm on Windows so I couldn't test Safari but according to this website: https://caniuse.com/?search=background-position-x
![image](https://user-images.githubusercontent.com/42903164/107472066-2264f080-6b23-11eb-9e30-40a73407c224.png)

Safari should be supported, but just to be safe if you have Safari maybe you can try out the changes.

elebitzero commented:
Hey, I remember I still have a free browserstack.com account so I can test the latest Safari (14) on Big Sur Mac Os.  Here are the screenshots of before and after. Looks the same as on Windows.   So, the only one we haven't checked yet is Safari on iOS.
**BEFORE**
![image](https://user-images.githubusercontent.com/42903164/107481774-98715380-6b33-11eb-90f3-b86edc66130f.png)
**AFTER**
![image](https://user-images.githubusercontent.com/42903164/107481790-9d360780-6b33-11eb-89e0-12ee7b7eb983.png)

elebitzero commented:
I just found out  browserstack.com  also lets me test on physical iOS devices remotely.  I just tried it out on an iPhone 12 running Safari, and it looks good. See pictures below.
**BEFORE**
![image](https://user-images.githubusercontent.com/42903164/107482658-da4ec980-6b34-11eb-81df-ba457b1dad89.png)
**AFTER**
![image](https://user-images.githubusercontent.com/42903164/107482675-df137d80-6b34-11eb-922d-a35f04541ca3.png)

uzair-ashraf commented:
Awesome!, I will add some change requests for the sake of complexity and possible time heavy operations where not needed

uzair-ashraf commented on src/data/characters.json:
```diff
@@ -1,192 +1,220 @@
+    "element": "Hydro",
+    "type": "character",
+    "rating": 5,
+    "percentX": 60
+  },
+  {
+    "name": "Diona",
+    "element": "Cryo",
+    "type": "character",
+    "rating": 4,
+    "percentX": 57
+  },
+  {
+    "name": "Zhongli",
+    "element": "Geo",
+    "type": "character",
+    "rating": 5,
+    "percentX": 58
+  },
+  {
+    "name": "Xinyan",
+    "element": "Pyro",
+    "type": "character",
+    "rating": 4,
+    "percentX": 60
+  },
+  {
+    "name": "Albedo",
+    "element": "Geo",
+    "type": "character",
+    "rating": 5,
+    "percentX": 56
+  },
+  {
+    "name": "Ganyu",
+    "element": "Cryo",
+    "type": "character",
+    "rating": 5,
+    "percentX": 57
+  }
```
So we can keep the percentX here for future banners when I create them automatically. However, we should have them on the items in the banner directly, I'll explain in the other files.

uzair-ashraf commented on src/components/wish-results.jsx:
```diff
@@ -2,10 +2,24 @@ import React, { Component } from 'react'
   isNewItem(key) {
     return !this.props.inventory[key]
   }
+  getPercentX(item) {
+    if (item.type === 'character') {
+      let matches = characters.filter(c => c.name === item.name);
+      if (matches.length === 1) {
+        return matches[0].percentX || 50;
+      } else {
+        console.log('Cannot find character '+item.name+' in characters.json');
+        return 50;
+      }
+    } else {
+      return 50;
+    }
+  }
```
I don't like the idea that every time we get a character in the banner, we have to iterate through a list of a bunch of characters to find the matching one. It would be much more efficient if we just add properties on characters in the banners.
You will have to update the schema validation objects to account for the percentX property

uzair-ashraf commented on src/components/wish-item.jsx:
```diff
@@ -4,14 +4,15 @@ import ReactStars from "react-rating-stars-component";
     style={{
-      backgroundImage: `url('${type === 'character' ? characterImages('./' + src).default : weaponImages('./' + src).default}')`
+      backgroundImage: `url('${type === 'character' ? characterImages('./' + src).default : weaponImages('./' + src).default}')`,
+      backgroundPositionX: characterPercentX+'%'
```
the backgroundPositionX doesn't effect weapons right?

elebitzero replied:
I kept the characters.json and updated the rest of the character data in the banners and the code in beginner's wish to have percentX and tested it again and it still looks good. Weapons are set to 50%, which is what is was before. Nevertheless, characterPercentX does set the horizontal offset for weapons so I renamed it to itemPercentX.

elebitzero commented:
I kept the characters.json and updated the rest of the character data in the banners and the code in beginner's wish to have percentX and tested it again and it still looks good. Weapons are set to 50%, which is what is was before. Nevertheless, characterPercentX does set the horizontal offset for weapons so I renamed it to itemPercentX.

elebitzero commented:
I also added a fix for Issue#67
